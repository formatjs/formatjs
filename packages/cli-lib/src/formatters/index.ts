import * as defaultFormatter from './default'
import {FormatFn, CompileFn} from './default'
import * as transifex from './transifex'
import * as smartling from './smartling'
import * as simple from './simple'
import * as lokalise from './lokalise'
import * as crowdin from './crowdin'
import {Comparator} from 'json-stable-stringify'
import {resolve} from 'path'
import {pathToFileURL} from 'url'

export interface Formatter {
  format: FormatFn
  compile: CompileFn
  compareMessages?: Comparator
}

export async function resolveBuiltinFormatter(
  format?: string | Formatter
): Promise<any> {
  if (!format) {
    return defaultFormatter
  }
  if (typeof format !== 'string') {
    return format
  }
  switch (format) {
    case 'transifex':
      return transifex
    case 'smartling':
      return smartling
    case 'simple':
      return simple
    case 'lokalise':
      return lokalise
    case 'crowdin':
      return crowdin
  }
  try {
    // eslint-disable-next-line import/dynamic-import-chunkname
    return import(pathToFileURL(resolve(process.cwd(), format)).href)
  } catch (e) {
    console.error(`Cannot resolve formatter ${format}`)
    throw e
  }
}
