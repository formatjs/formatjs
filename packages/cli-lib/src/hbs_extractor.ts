import {transform} from 'ember-template-recast'

function extractText(node: any, fileName: any, options: any) {
  if (['format-message', 'formatMessage'].includes(node.path.original)) {
    let message = node.params[0]?.original
    let desc = node.params[1]?.original?.trim().replace(/\s+/gm, ' ')
    let defaultMessage = message?.trim().replace(/\s+/gm, ' ')
    // overrideIdFn needs to receive defaultMessage/description with already trimmed whitespaces
    // so the resulting id is the same if messages have the same content but have different indentation
    // id has to be generated the same way in soxhub/ember-formatjs/replace-key-transform !!!!!!!!!!!!!!!!!!!!!!
    let id = options.overrideIdFn(undefined, defaultMessage, desc, fileName)
    options.onMsgExtracted(undefined, {
      id: id,
      defaultMessage: defaultMessage,
      description: desc,
    })
  }
}

export function parseFile(source: string, fileName: string, options: any) {
  let visitor = function () {
    return {
      MustacheStatement(node: any) {
        extractText(node, fileName, options)
      },
      SubExpression(node: any) {
        extractText(node, fileName, options)
      },
    }
  }

  transform(source, visitor)
}
