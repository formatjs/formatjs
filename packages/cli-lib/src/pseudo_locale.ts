import {
  parse,
  MessageFormatElement,
  TYPE,
  isLiteralElement,
  isPluralElement,
  isSelectElement,
  isTagElement,
  LiteralElement,
} from '@formatjs/icu-messageformat-parser'

function forEachLiteralElement(
  ast: MessageFormatElement[],
  fn: (el: LiteralElement) => void
): void {
  ast.forEach(el => {
    if (isLiteralElement(el)) {
      fn(el)
    } else if (isPluralElement(el) || isSelectElement(el)) {
      for (const opt of Object.values(el.options)) {
        forEachLiteralElement(opt.value, fn)
      }
    } else if (isTagElement(el)) {
      forEachLiteralElement(el.children, fn)
    }
  })
}

export function generateXXLS(
  msg: string | MessageFormatElement[]
): MessageFormatElement[] {
  const ast = typeof msg === 'string' ? parse(msg) : msg
  const lastChunk = ast[ast.length - 1]
  if (lastChunk && isLiteralElement(lastChunk)) {
    lastChunk.value += 'SSSSSSSSSSSSSSSSSSSSSSSSS'
    return ast
  }
  return [...ast, {type: TYPE.literal, value: 'SSSSSSSSSSSSSSSSSSSSSSSSS'}]
}

export function generateXXAC(
  msg: string | MessageFormatElement[]
): MessageFormatElement[] {
  const ast = typeof msg === 'string' ? parse(msg) : msg
  forEachLiteralElement(ast, el => {
    el.value = el.value.toUpperCase()
  })
  return ast
}

export function generateXXHA(
  msg: string | MessageFormatElement[]
): MessageFormatElement[] {
  const ast = typeof msg === 'string' ? parse(msg) : msg
  const [firstChunk, ...rest] = ast
  if (firstChunk && isLiteralElement(firstChunk)) {
    firstChunk.value = '[javascript]' + firstChunk.value
    return [firstChunk, ...rest]
  }
  return [{type: TYPE.literal, value: '[javascript]'}, ...ast]
}

type _TupleOf<T, N extends number, R extends unknown[]> = R['length'] extends N
  ? R
  : _TupleOf<T, N, [T, ...R]>

type Tuple<T, N extends number> = N extends N
  ? number extends N
    ? T[]
    : _TupleOf<T, N, []>
  : never

type PseudoLocaleTransformMap = {
  caps: Tuple<number, 26>
  small: Tuple<number, 26>
}

const ACCENTED_MAP: PseudoLocaleTransformMap = {
  // ȦƁƇḒḖƑƓĦĪĴĶĿḾȠǾƤɊŘŞŦŬṼẆẊẎẐ
  // prettier-ignore
  "caps": [550, 385, 391, 7698, 7702, 401, 403, 294, 298, 308, 310, 319, 7742, 544, 510, 420, 586, 344, 350, 358, 364, 7804, 7814, 7818, 7822, 7824],
  // ȧƀƈḓḗƒɠħīĵķŀḿƞǿƥɋřşŧŭṽẇẋẏẑ
  // prettier-ignore
  "small": [551, 384, 392, 7699, 7703, 402, 608, 295, 299, 309, 311, 320, 7743, 414, 511, 421, 587, 345, 351, 359, 365, 7805, 7815, 7819, 7823, 7825],
}

const FLIPPED_MAP: PseudoLocaleTransformMap = {
  // ∀ԐↃᗡƎℲ⅁HIſӼ⅂WNOԀÒᴚS⊥∩ɅMX⅄Z
  // prettier-ignore
  "caps": [8704, 1296, 8579, 5601, 398, 8498, 8513, 72, 73, 383, 1276, 8514, 87, 78, 79, 1280, 210, 7450, 83, 8869, 8745, 581, 77, 88, 8516, 90],
  // ɐqɔpǝɟƃɥıɾʞʅɯuodbɹsʇnʌʍxʎz
  // prettier-ignore
  "small": [592, 113, 596, 112, 477, 607, 387, 613, 305, 638, 670, 645, 623, 117, 111, 100, 98, 633, 115, 647, 110, 652, 653, 120, 654, 122],
}

/**
 * Based on: https://hg.mozilla.org/mozilla-central/file/a1f74e8c8fb72390d22054d6b00c28b1a32f6c43/intl/l10n/L10nRegistry.jsm#l425
 */
function transformString(
  map: PseudoLocaleTransformMap,
  elongate = false,
  msg: string
) {
  return msg.replace(/[a-z]/gi, ch => {
    const cc = ch.charCodeAt(0)
    if (cc >= 97 && cc <= 122) {
      const newChar = String.fromCodePoint(map.small[cc - 97])
      // duplicate "a", "e", "o" and "u" to emulate ~30% longer text
      if (elongate && (cc === 97 || cc === 101 || cc === 111 || cc === 117)) {
        return newChar + newChar
      }
      return newChar
    }
    if (cc >= 65 && cc <= 90) {
      return String.fromCodePoint(map.caps[cc - 65])
    }
    return ch
  })
}

/**
 * accented - Ȧȧƈƈḗḗƞŧḗḗḓ Ḗḗƞɠŀīīşħ
 * --------------------------------
 *
 * This locale replaces all Latin characters with their accented equivalents, and duplicates some
 * vowels to create roughly 30% longer strings. Strings are wrapped in markers (square brackets),
 * which help with detecting truncation.
 */
export function generateENXA(
  msg: string | MessageFormatElement[]
): MessageFormatElement[] {
  const ast = typeof msg === 'string' ? parse(msg) : msg
  forEachLiteralElement(ast, el => {
    el.value = transformString(ACCENTED_MAP, true, el.value)
  })
  return [
    {type: TYPE.literal, value: '['},
    ...ast,
    {type: TYPE.literal, value: ']'},
  ]
}

/**
 * bidi - ɥsıʅƃuƎ ıpıԐ
 * -------------------
 *
 * This strategy replaces all Latin characters with their 180 degree rotated versions and enforces
 * right to left text flow using Unicode UAX#9 Explicit Directional Embeddings. In this mode, the UI
 * directionality will also be set to right-to-left.
 */
export function generateENXB(
  msg: string | MessageFormatElement[]
): MessageFormatElement[] {
  const ast = typeof msg === 'string' ? parse(msg) : msg
  forEachLiteralElement(ast, el => {
    el.value = transformString(FLIPPED_MAP, false, el.value)
  })
  return [
    {type: TYPE.literal, value: '\u202e'},
    ...ast,
    {type: TYPE.literal, value: '\u202c'},
  ]
}
