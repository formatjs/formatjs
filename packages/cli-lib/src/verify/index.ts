import {basename} from 'path'
import {debug} from '../console_utils.js'
import {checkMissingKeys} from './checkMissingKeys.js'
import {readJSON} from 'fs-extra'
import {checkStructuralEquality} from './checkStructuralEquality.js'
export interface VerifyOpts {
  sourceLocale: string
  missingKeys: boolean
  structuralEquality: boolean
  ignore?: string[]
}

export async function verify(
  files: string[],
  {sourceLocale, missingKeys, structuralEquality}: VerifyOpts
): Promise<void> {
  debug('Checking translation files:')
  files.forEach(fn => debug(fn))

  const translationFilesContents = (
    await Promise.all(
      files.map(async fn => [basename(fn, '.json'), await readJSON(fn)])
    )
  ).reduce<Record<string, object>>((all, [locale, content]) => {
    all[locale] = content
    return all
  }, {})
  debug('Verifying files:', files)

  let exitCode = 0

  if (
    missingKeys &&
    !(await checkMissingKeys(translationFilesContents, sourceLocale))
  ) {
    exitCode = 1
  }

  if (
    structuralEquality &&
    !(await checkStructuralEquality(translationFilesContents, sourceLocale))
  ) {
    exitCode = 1
  }

  process.exit(exitCode)
}
