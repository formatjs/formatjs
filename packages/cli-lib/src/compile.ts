import {parse, MessageFormatElement} from '@formatjs/icu-messageformat-parser'
import {outputFile, readJSON} from 'fs-extra'
import stringify from 'json-stable-stringify'
import {debug, warn, writeStdout} from './console_utils'
import {resolveBuiltinFormatter, Formatter} from './formatters'
import {
  generateXXAC,
  generateXXLS,
  generateXXHA,
  generateENXA,
  generateENXB,
} from './pseudo_locale'

export type CompileFn = (msgs: any) => Record<string, string>

export type PseudoLocale = 'xx-LS' | 'xx-AC' | 'xx-HA' | 'en-XA' | 'en-XB'

export interface CompileCLIOpts extends Opts {
  /**
   * The target file that contains compiled messages.
   */
  outFile?: string
}
export interface Opts {
  /**
   * Whether to compile message into AST instead of just string
   */
  ast?: boolean
  /**
   * Whether to continue compiling messages after encountering an error.
   * Any keys with errors will not be included in the output file.
   */
  skipErrors?: boolean
  /**
   * Path to a formatter file that converts <translation_files> to
   * `Record<string, string>` so we can compile.
   */
  format?: string | Formatter
  /**
   * Whether to compile to pseudo locale
   */
  pseudoLocale?: PseudoLocale
}

/**
 * Aggregate `inputFiles` into a single JSON blob and compile.
 * Also checks for conflicting IDs.
 * Then returns the serialized result as a `string` since key order
 * makes a difference in some vendor.
 * @param inputFiles Input files
 * @param opts Options
 * @returns serialized result in string format
 */
export async function compile(inputFiles: string[], opts: Opts = {}) {
  debug('Compiling files:', inputFiles)
  const {ast, format, pseudoLocale, skipErrors} = opts
  const formatter = await resolveBuiltinFormatter(format)

  const messages: Record<string, string> = {}
  const messageAsts: Record<string, MessageFormatElement[]> = {}
  const idsWithFileName: Record<string, string> = {}
  const compiledFiles = await Promise.all(
    inputFiles.map(f => readJSON(f).then(formatter.compile))
  )
  debug('Compiled files:', compiledFiles)
  for (let i = 0; i < inputFiles.length; i++) {
    const inputFile = inputFiles[i]
    debug('Processing file:', inputFile)
    const compiled = compiledFiles[i]
    for (const id in compiled) {
      if (messages[id] && messages[id] !== compiled[id]) {
        throw new Error(`Conflicting ID "${id}" with different translation found in these 2 files:
ID: ${id}
Message from ${idsWithFileName[id]}: ${messages[id]}
Message from ${inputFile}: ${compiled[id]}
`)
      }
      try {
        const msgAst = parse(compiled[id])
        messages[id] = compiled[id]
        switch (pseudoLocale) {
          case 'xx-LS':
            messageAsts[id] = generateXXLS(msgAst)
            break
          case 'xx-AC':
            messageAsts[id] = generateXXAC(msgAst)
            break
          case 'xx-HA':
            messageAsts[id] = generateXXHA(msgAst)
            break
          case 'en-XA':
            messageAsts[id] = generateENXA(msgAst)
            break
          case 'en-XB':
            messageAsts[id] = generateENXB(msgAst)
            break
          default:
            messageAsts[id] = msgAst
            break
        }
        idsWithFileName[id] = inputFile
      } catch (e) {
        warn(
          'Error validating message "%s" with ID "%s" in file "%s"',
          compiled[id],
          id,
          inputFile
        )
        if (!skipErrors) {
          throw e
        }
      }
    }
  }

  return stringify(ast ? messageAsts : messages, {
    space: 2,
    cmp: formatter.compareMessages || undefined,
  })
}

/**
 * Aggregate `inputFiles` into a single JSON blob and compile.
 * Also checks for conflicting IDs and write output to `outFile`.
 * @param inputFiles Input files
 * @param compileOpts options
 * @returns A `Promise` that resolves if file was written successfully
 */
export default async function compileAndWrite(
  inputFiles: string[],
  compileOpts: CompileCLIOpts = {}
) {
  const {outFile, ...opts} = compileOpts
  const serializedResult = await compile(inputFiles, opts)
  if (outFile) {
    debug('Writing output file:', outFile)
    return outputFile(outFile, serializedResult)
  }
  await writeStdout(serializedResult)
  await writeStdout('\n')
}
