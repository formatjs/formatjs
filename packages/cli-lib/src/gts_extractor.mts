import {parseFile as parseHbsFile} from './hbs_extractor.mjs'
import {parseScript} from './parse_script.js'
import {Preprocessor} from 'content-tag'
let p = new Preprocessor()

export function parseFile(source: string, fileName: string, options: any) {
  const scriptParseFn = parseScript(options, fileName)
  const transformedSource = p.process(source, {filename: fileName})

  scriptParseFn(transformedSource)

  // extract template from transformed source to then run through hbs processor
  const parseResult = p.parse(source, {filename: fileName})

  for (let parsed of parseResult) {
    parseHbsFile(parsed.contents, fileName, options)
  }
}
