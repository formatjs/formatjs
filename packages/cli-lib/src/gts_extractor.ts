// @ts-ignore
import {transform, parseTemplates} from 'ember-template-tag'
import {parseFile as parseHbsFile} from './hbs_extractor'
import {parseScript} from './parse_script'

export function parseFile(source: string, fileName: string, options: any) {
  const scriptParseFn = parseScript(options, fileName)
  const {output: transformedSource} = transform({
    input: source,
    relativePath: '',
  })

  scriptParseFn(transformedSource)

  // extract template from transformed source to then run through hbs processor
  const [templateSource] = parseTemplates(source, '')

  parseHbsFile(templateSource.contents, fileName, options)
}
