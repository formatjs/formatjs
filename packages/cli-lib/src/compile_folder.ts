import {outputFile} from 'fs-extra'
import {basename, join} from 'path'
import {Opts, compile} from './compile.js'
export default async function compileFolder(
  files: string[],
  outFolder: string,
  opts: Opts = {}
): Promise<void[]> {
  const results = await Promise.all(files.map(f => compile([f], opts)))
  const outFiles = files.map(f => join(outFolder, basename(f)))

  return Promise.all(
    outFiles.map((outFile, i) => outputFile(outFile, results[i]))
  )
}
