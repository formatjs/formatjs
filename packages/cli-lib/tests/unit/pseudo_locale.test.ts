import {generateXXLS} from '../../src/pseudo_locale'

describe('pseudo-locale: xx-LS', () => {
  it('works with messages that ends with a tag', () => {
    expect(generateXXLS('Foo <a>bar</a>')).toMatchSnapshot()
  })

  it('works with message that does not contain literal element', () => {
    expect(generateXXLS('{foo}')).toMatchSnapshot()
  })
})

describe('pseudo-locale: en-HA', () => {
  it('works with message that does not contain literal element', () => {
    expect(generateXXLS('{foo}')).toMatchSnapshot()
  })
})
