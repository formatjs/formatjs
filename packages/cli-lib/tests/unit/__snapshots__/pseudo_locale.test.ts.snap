// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pseudo-locale: en-HA works with message that does not contain literal element 1`] = `
Array [
  Object {
    "type": 1,
    "value": "foo",
  },
  Object {
    "type": 0,
    "value": "SSSSSSSSSSSSSSSSSSSSSSSSS",
  },
]
`;

exports[`pseudo-locale: xx-LS works with message that does not contain literal element 1`] = `
Array [
  Object {
    "type": 1,
    "value": "foo",
  },
  Object {
    "type": 0,
    "value": "SSSSSSSSSSSSSSSSSSSSSSSSS",
  },
]
`;

exports[`pseudo-locale: xx-LS works with messages that ends with a tag 1`] = `
Array [
  Object {
    "type": 0,
    "value": "Foo ",
  },
  Object {
    "children": Array [
      Object {
        "type": 0,
        "value": "bar",
      },
    ],
    "type": 8,
    "value": "a",
  },
  Object {
    "type": 0,
    "value": "SSSSSSSSSSSSSSSSSSSSSSSSS",
  },
]
`;
